diff --git a/s3_storage_provider.py b/s3_storage_provider.py
index cbf4000..2285fd6 100644
--- a/s3_storage_provider.py
+++ b/s3_storage_provider.py
@@ -57,7 +57,6 @@ class S3StorageProviderBackend(StorageProvider):
 
     def __init__(self, hs, config):
         self._module_api: ModuleApi = hs.get_module_api()
-        self.cache_directory = hs.config.media.media_store_path
         self.bucket = config["bucket"]
         self.prefix = config["prefix"]
         # A dictionary of extra arguments for uploading files.
@@ -120,13 +119,13 @@ class S3StorageProviderBackend(StorageProvider):
                 self._s3_client = s3 = b3_session.client("s3", **self.api_kwargs)
             return s3
 
-    async def store_file(self, path, file_info):
+    async def store_file(self, path, file_info, file_path):
         """See StorageProvider.store_file"""
 
         return await self._module_api.defer_to_threadpool(
             self._s3_pool,
             self._get_s3_client().upload_file,
-            Filename=os.path.join(self.cache_directory, path),
+            Filename=file_path,
             Bucket=self.bucket,
             Key=self.prefix + path,
             ExtraArgs=self.extra_args,
